### PR TITLE
[FIX] Fixes issue #29

### PR DESCRIPTION
**Pull request by JordiBForgeFlow**
This fix would solve issue #29. 

The following corrections are introduced:
- A user can list the po lines in menu "On Purchase Order Lines" when the PO line has been partially invoiced. 
- The PO that has been created with invoice control 'On Purchase Order Lines' will only change to status "Done" from "Purchase Order" when all the PO lines have been fully invoiced, and the invoices, approved.  In order to do that, the field 'invoiced' in the PO is only flagged when all PO lines have been fully invoiced AND for those lines that were fully invoiced, there's no invoice in draft or cancel status. 

Regards,
Jordi.

### CODE REVIEW

**Comment by pedrobaeza** [comment]
Hi, Jordi,

Thanks for the improvements. The only point I see conflictive is the one about allow to reinvoice cancel invoiced quantities. I prefer not to do it, but force to remove the invoice line or the full invoice, because this can produce that you reinvoice, and then back to draft cancelled invoice, invoicing as result two times the line.

**Comment by JordiBForgeFlow** [comment]
@pedrobaeza Agreed. I just corrected as per your comment.  Thanks for your review!

File: purchase_partial_invoicing/purchase.py [hunk 1]
```diff
@@ -33,6 +33,25 @@ def _invoiced_qty(self, cursor, user, ids, name, arg, context=None):
             res[line.id] = invoiced_qty
         return res
 
+    def _fully_invoiced(self, cursor, user, ids, name, arg, context=None):
+        res = {}
+        for line in self.browse(cursor, user, ids, context=context):
+            res[line.id] = False
```
Comment (pedrobaeza) [inline purchase_partial_invoicing/purchase.py]: Better to put just this:

```
res[line.id] = line.invoiced_qty == line.product_qty
```

File: purchase_partial_invoicing/purchase.py [hunk 2]
```diff
@@ -33,6 +33,25 @@ def _invoiced_qty(self, cursor, user, ids, name, arg, context=None):
...
+
+    def _all_invoices_approved(self, cursor, user, ids, name, arg, context=None):
+        res = {}
+        for line in self.browse(cursor, user, ids, context=context):
+            res[line.id] = False
+            if line.fully_invoiced:
+                for invoice_line in line.invoice_lines:
+                    if invoice_line.invoice_id.state not in ['draft',
+                                                             'cancel']:
+                        res[line.id] = True
```
Comment (pedrobaeza) [inline purchase_partial_invoicing/purchase.py]: This shouldn't work as the meaning of the field, being True if all the invoices are approved. It just needs one.

Comment (yvaucher) [inline purchase_partial_invoicing/purchase.py]: As @pedrobaeza said, you should inverse the logic and break if any invoice is in `draft` or `cancel` state

**Comment by JordiBForgeFlow** [comment]
Corrected as sugested, @pedrobaeza and @yvaucher.

**Comment by pedrobaeza** [comment]
:+1:

File: purchase_partial_invoicing/purchase.py [hunk 3]
```diff
@@ -33,6 +33,23 @@ def _invoiced_qty(self, cursor, user, ids, name, arg, context=None):
             res[line.id] = invoiced_qty
         return res
 
+    def _fully_invoiced(self, cursor, user, ids, name, arg, context=None):
+        res = {}
+        for line in self.browse(cursor, user, ids, context=context):
+            res[line.id] = line.invoiced_qty == line.product_qty
+        return res
+
+    def _all_invoices_approved(self, cursor, user, ids, name, arg, context=None):
```
Comment (yvaucher) [inline purchase_partial_invoicing/purchase.py]: ./purchase_partial_invoicing/purchase.py:42:80: E501 line too long (81 > 79 characters)

File: purchase_partial_invoicing/purchase.py [hunk 4]
```diff
@@ -40,4 +57,45 @@ def _invoiced_qty(self, cursor, user, ids, name, arg, context=None):
...
+
+
+class purchase_order(orm.Model):
+
+    _inherit = 'purchase.order'
+
+    def _invoiced(self, cursor, user, ids, name, arg, context=None):
+        res = {}
+        for purchase in self.browse(cursor, user, ids, context=context):
+            res[purchase.id] = all(line.all_invoices_approved for line in purchase.order_line)
```
Comment (yvaucher) [inline purchase_partial_invoicing/purchase.py]: ./purchase_partial_invoicing/purchase.py:78:80: E501 line too long (94 > 79 characters)

File: purchase_partial_invoicing/purchase.py [hunk 5]
```diff
@@ -40,4 +57,45 @@ def _invoiced_qty(self, cursor, user, ids, name, arg, context=None):
...
+                                    help="It indicates that an invoice has "
+                                         "been validated"),
+    }
+
+
+class account_invoice(orm.Model):
+    _inherit = 'account.invoice'
+
+    def invoice_validate(self, cr, uid, ids, context=None):
+        res = super(account_invoice, self).invoice_validate(cr, uid, ids, context=context)
```
Comment (yvaucher) [inline purchase_partial_invoicing/purchase.py]: ./purchase_partial_invoicing/purchase.py:93:80: E501 line too long (90 > 79 characters)

File: purchase_partial_invoicing/purchase.py [hunk 6]
```diff
@@ -40,4 +57,45 @@ def _invoiced_qty(self, cursor, user, ids, name, arg, context=None):
...
+        purchase_order_obj = self.pool.get('purchase.order')
+        po_ids = purchase_order_obj.search(
+            cr, uid, [('invoice_ids', 'in', ids)], context=context)
+        for purchase_order in purchase_order_obj.browse(cr, uid, po_ids, context=context):
```
Comment (yvaucher) [inline purchase_partial_invoicing/purchase.py]: ./purchase_partial_invoicing/purchase.py:97:80: E501 line too long (90 > 79 characters)

**Comment by yvaucher** [comment]
PEP8 to fix

**Comment by JordiBForgeFlow** [comment]
@yvaucher Fixed PEP8. And finally saw how to set the right margin in pyCharm to 79 lines.... :)

**Comment by lepistone** [comment]
Note: for github to recognize the commit message and close the issue automatically, you probably need to say something like `Fixes #29` without the `issue` word in between. https://help.github.com/articles/closing-issues-via-commit-messages

Otherwise, we'll just close the issue manually.

**Comment by JordiBForgeFlow** [comment]
@lepistone Thanks. Done.

**Comment by lepistone** [comment]
@jbeficent in the last commit to removed a newline, introducing a lint error. It actually is not a good idea to introduce an empty commit just to say "fixes #29". Best thing would be to rebase, squash the fixups and change a commit message, but I'm happy to just close the issues manually instead.

Don't worry about the lint problems in other modules: they should go away in the next build.

**Comment by JordiBForgeFlow** [comment]
@lepistone. Ok. I will keep in mind your advice. Close this one manually if it's ok with you. Thanks!!

**Comment by lepistone** [comment]
No problem, I'll close the issue when merging (and squash the commits, probably).

On the other hand, the red builds because of the newline removed still stand and should be fixed, please.

@eLBati, you're the original author. Do you approve this?

**Comment by eLBati** [comment]
Thanks @jbeficent 

> If an invoice from the PO line has been cancelled, it does not count as invoiced quantity for that PO line

How do you realize this? I see no changes in `_invoiced_qty` method.

**Comment by JordiBForgeFlow** [comment]
Hi @elbati, @pedrobaeza recommeded in his review not no consider this. See hs firdt review above. I will change the description of the issue

**Comment by eLBati** [comment]
Ok @jbeficent thanks.
Just a PEP8 remark:
`./purchase_partial_invoicing/wizard/po_line_invoice.py:108:6: W292 no newline at end of file`
Then :+1:

**Comment by adrienpeiffer** [comment]
:+1: If travis becomes well

**Comment by JordiBForgeFlow** [comment]
@adrienpeiffer @eLBati  Fixed PEP8 remark. Thanks!

**Comment by eLBati** [comment]
Travis failed due to github timeout. Can it be restarted?

**Comment by yvaucher** [comment]
travis restarted

**Comment by yvaucher** [comment]
Travis error on `purchase_landed_costs` is unrelated

**Pull request merged**